### DISCUSSION TITLE
New component: lg-slider api

### DISCUSSION THREAD

**Discussion by elenagarrone** [Legal-and-General/canopy#182] [General]
# Slider component

I need to build a slider component for our project and I would like to have a discussion about what the component API should look like.

I think we would need a basic component - simple slider like the one Material has (https://material.angular.io/components/slider/overview) that can have a more complex version like the one in the design below that allows you to divide the current value into different `sections`.

![Screenshot 2021-01-27 at 14 18 22](https://user-images.githubusercontent.com/8397116/106003877-89f25900-60aa-11eb-9846-7b5f361a6286.png)

For the basic version the API would look like this:
```
<lg-slider></lg-slider>
```

with the following properties:
| Name        | Description |
| ----------- | ----------- |
| @Input() max: number | The maximum value that the slider can have |
| @Input() min: number | The minimum value that the slider can have |
| @Input() value: number | The current value of the slider thumb |
| @Output() change: number | The current value emitted on change |

Do we need to pass colours too or do we have defaulted ones? If we pass a colour do we have: `trackColor` and `valueColor`?
Would we use `colour` or `color` as an input?

For the more advanced version I would probably add a new component?
```
<lg-slider>
    <lg-slider-section></lg-slider-section>
</slider>
```

the slider section would have the following inputs:
| Name        | Description |
| ----------- | ----------- |
| @Input() percentage: number | The percentage of value to "cover" (e.g. the darkest blue section in the design) |

Same here, do we need a colour input?

I haven't tried this at all and I have never created a slider before so any idea/suggestion you have please let me know.

**Comment by pauleustice** [comment]
Sounds like a reasonable approach. I think the `slider-section` would need a concept of `range` rather than a `percentage` input, as you need to specify a start and end point. This might also be two inputs. e.g.

```
<lg-slider-segment start="10" end="35"></lg-slider-segment>
```

If start is omitted, it could default to 0, and end could default to 100.

**Reply by pauleustice** [reply]
I'd also favour the ability to customise colours. It could accept colour keywords representing the Canopy colours, or any CSS format.

On the spelling, well I'd say 'colour' as we're a British organisation but that may be contentious :)

**Reply by Devaney** [reply]
Agree it sounds good 👍  may have to consider adding a `step` input as well so we can adjust the decimal places on the value emitted.
`<lg-slider-segment start="0" end="100" type="range" step="1"></lg-slider-segment>` < can output 1

`<lg-slider-segment start="0" end="100" type="range" step="0.01"></lg-slider-segment>` < can output 1.01

**Reply by owensgit** [reply]
> I think the slider-section would need a concept of range rather than a percentage input

I agree with this, I like the idea of having `start` and `end` inputs.

> I'd also favour the ability to customise colours. It could accept colour keywords representing the Canopy colours, or any CSS format.

I think I'd favour colour keywords, as they would adhere to the brand colours and ensure we only render colours that the designers have signed off for use on the slider - we tend to be quite prescriptive about the use of colours across our components - for example with the variants. 

> may have to consider adding a step input

Good idea - although I wonder if `increment` or `incrementBy` is a better name? As the value would increment by that amount on each slider change event.

**Reply by andy-polhill** [reply]
On the colour point I had been contemplating a background colour directive `lgBackgroundColor="cherry-red`.
I think it could provide quite a lot of flexibility, but also be open to abuse.

I agree with @Devaney on the [step](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input#attr-step) property

**Comment by owensgit** [comment]
Looks good to me 👍  I also agree with the points raised by @pauleustice  and @Devaney. 

> Do we need to pass colours too or do we have defaulted ones?

I think it would be good to have some defaults that can be overridden the inputs.

> If we pass a colour do we have: `trackColor` and `valueColor`? Would we use colour or color as an input?

I always go with "colour", as it's the British spelling, although, we have used "color" in the CSS vars 😬 So I wouldn't mind so much if we wanted to keep it consistent.

In terms of the input names, perhaps they could be `trackColour` and `valueControlColour`, I think adding the word "control" makes it clear it's to style the colour of the circle that you drag (if that is the intention).

**Reply by andy-polhill** [reply]
I must admit I tend to go for `color` throughout just for consistency.

I would check with @alexcanning about the colour, It could be that it's always the same colour. The sections might also end up being the same.

**Reply by elenagarrone** [reply]
> I must admit I tend to go for `color` throughout just for consistency.
> 
> I would check with @alexcanning about the colour, It could be that it's always the same colour. The sections might also end up being the same.

I doubt the section would always be the same colour - you never know how many sections a user might be using 🤔 

I think I agree with you on `color`; in code I'm used to write it the non-british way so it makes a bit more sense for it to be consistent 🤔

**Reply by alexcanning** [reply]
I don't think we can say that it'll be the same colour every time - I think the ambition here is to have those sections relate to something colour-coded elsewhere on the page. I'd prefer if we used the documentation and brand guidelines to suggest the correct order, but leave it up to the judgment of the team implementing the control.
